File: apps/web/src/data/challenge.ts
// Types for the static connections-demo artifact (public/data/challenge.v1.json).
//
// These mirror the real credits schema produced by the Discogs ingestion
// pipeline in packages/catalog (see discogs/parquet.py CREDIT_SCHEMA and
// discogs/releases.py). The artifact is real Discogs data for a small, curated
// subset of releases (see packages/catalog discogs/demo_challenge.py and ADR
// 0012) — not the full private seed, and not yet the CC0 monthly-dump-derived
// dataset that Milestone 8 will eventually produce under this same shape.

/** One credited contribution. Faithful to packages/catalog CREDIT_SCHEMA. */
export interface Credit {
  snapshot_date: string;
  release_id: number;
  track_index: number | null;
  track_path: string | null;
  track_position: string | null;
  track_title: string | null;
  /** release_artist | release_credit | track_artist | track_credit */
  credit_scope:
    "release_artist" | "release_credit" | "track_artist" | "track_credit";
  /** Discogs linked artist id (PAN). null when the contributor is not linked. */
  artist_id: number | null;
  /** Credited name as given in the source. */
  name: string;
  /** Artist Name Variation — display override; kept separate from artist_id. */
  anv: string | null;
  join_text: string | null;
  /** Original, un-normalized role text. */
  role_text: string | null;
  credited_tracks_text: string | null;
  /** true when artist_id resolves to a linked artist. */
  is_linked: boolean;
  /** true only for linked artists that may act as graph nodes. */
  playable_identity: boolean;
}

/** A cover-art image, hotlinked from Discogs' own CDN — never downloaded or rehosted. */
export interface ReleaseImage {
  uri: string;
  uri150: string;
  width: number;
  height: number;
}

/** A release plus its credit rows. */
export interface Release {
  snapshot_date: string;
  release_id: number;
  status: string;
  title: string;
  country: string | null;
  released: string | null;
  master_id: number | null;
  master_is_main_release: boolean | null;
  data_quality: string | null;
  source_url: string;
  images: ReleaseImage[];
  credits: Credit[];
}

/** A playable identity (linked artist only). */
export interface Artist {
  artist_id: number;
  name: string;
}

/** One step of a path: two artists co-credited on the same release. */
export interface Hop {
  release_id: number;
  artist_a_id: number;
  artist_b_id: number;
}

/** A curated, documented path between two artists. */
export interface Path {
  id: string;
  label: string;
  description: string;
  from_artist_id: number;
  to_artist_id: number;
  hops: Hop[];
}

export interface Provenance {
  source: string;
  license: string;
  snapshot_date: string;
  source_url: string;
  generated_by: string;
  catalog_parser_version: string;
  catalog_schema_version: number;
  note: string;
}

export interface Challenge {
  schema_version: number;
  provenance: Provenance;
  releases: Release[];
  artists: Artist[];
  paths: Path[];
}

// --- Challenge v2 (album-centered) -----------------------------------------
// Mirrors packages/graph-core's challenge.py output; see
// data/contracts/challenge-v2.md for the full field-by-field contract.
// Albums are the entry points here; releases are demoted to evidence beneath
// the hops that justify each connection.

/** An album (a matched master or release) — the top-level browsing unit. */
export interface AlbumV2 {
  id: string;
  master_id: number | null;
  main_release_id: number;
  title: string;
  artist_id: number;
  artist: string;
  year: number | null;
  /** Presentational only, hotlinked from Discogs — never load-bearing evidence. */
  cover_image: ReleaseImage | null;
}

/** A release demoted to evidence: same fields as Release, minus images, with
 * credits filtered to only the linked artists that are hop endpoints. */
export interface EvidenceRelease {
  snapshot_date: string;
  release_id: number;
  status: string;
  title: string;
  country: string | null;
  released: string | null;
  master_id: number | null;
  master_is_main_release: boolean | null;
  data_quality: string | null;
  source_url: string;
  credits: Credit[];
}

/** A curated, documented path between two albums' artists. */
export interface PathV2 {
  id: string;
  label: string;
  description: string;
  from_album_id: string;
  to_album_id: string;
  from_artist_id: number;
  to_artist_id: number;
  hops: Hop[];
}

export interface ProvenanceV2 {
  source: string;
  license: string;
  snapshot_date: string;
  generated_by: string;
  graph_core_version: string;
  note: string;
}

export interface ChallengeV2 {
  schema_version: number;
  provenance: ProvenanceV2;
  albums: AlbumV2[];
  artists: Artist[];
  paths: PathV2[];
  releases: EvidenceRelease[];
}

File: packages/contracts/src/networked_players_contracts/challenge.py
"""Canonical, dependency-free validation for the challenge.v2 artifact.

challenge.v2 (`apps/web/public/data/challenge.v2.json`) is the album-centered,
evidence-preserving static challenge: an album -> artist -> album documented
credit path over a bounded one-hop working set (ADR 0012). Generation-time
validation lives in
`packages/graph-core/.../challenge.py::validate_challenge`, which delegates
to `challenge_failures` here so the two can never drift -- same pattern as
`record_routes.py`'s `validate_record_routes_artifact` ->
`record_routes_failures` (ADR 0046). `CHALLENGE_SCHEMA_VERSION` here must
agree with graph-core's own constant of the same name (each layer keeps its
own copy, same convention as every other artifact pair in this package).

Reuses `.rounds`'s `_seed_key_paths`/`_privacy_failures` for the no-leaked-
`seed`-key and forbidden-substring/phrase scan, rather than a second copy --
the same protection every other real artifact in this package already gets.

Pure Python (no lxml/pyarrow/duckdb), safe for the Pi fleet and the web build.
"""

from __future__ import annotations

from typing import Any

from .rounds import _privacy_failures, _seed_key_paths

CHALLENGE_SCHEMA_VERSION = 2

_TOP_LEVEL_KEYS = frozenset(
    {"schema_version", "provenance", "albums", "artists", "paths", "releases"}
)
# Per data/contracts/challenge-v2.md (RELEASE_SCHEMA minus `images`) plus the
# `credits` evidence array the builder adds. Must agree with graph-core's own
# `_RELEASE_KEYS` of the same name.
_RELEASE_KEYS = frozenset(
    {
        "snapshot_date",
        "release_id",
        "status",
        "title",
        "country",
        "released",
        "master_id",
        "master_is_main_release",
        "data_quality",
        "source_url",
        "credits",
    }
)
_PROVENANCE_REQUIRED = ("source", "license", "snapshot_date", "generated_by", "graph_core_version")


def challenge_failures(artifact: Any, catalog: Any | None = None) -> list[str]:
    """Return every contract failure in a challenge.v2 artifact.

    If `catalog` is given and the artifact's own `provenance.catalog_version`
    is not `None`, it must agree with `catalog["catalog_version"]`. A
    hand-written `{artist,title}` query list legitimately has no catalog to
    agree with (its own provenance note documents this) -- a `None`
    `catalog_version` is never itself a failure, with or without a `catalog`
    argument."""
    if not isinstance(artifact, dict):
        return ["challenge artifact must be an object"]

    failures: list[str] = []
    if set(artifact) != _TOP_LEVEL_KEYS:
        failures.append(f"unexpected top-level keys: {sorted(artifact)}")
    if artifact.get("schema_version") != CHALLENGE_SCHEMA_VERSION:
        failures.append(f"schema_version must be {CHALLENGE_SCHEMA_VERSION}")

    provenance = artifact.get("provenance")
    if not isinstance(provenance, dict):
        failures.append("provenance must be an object")
        provenance = {}
    else:
        for field_name in _PROVENANCE_REQUIRED:
            if not provenance.get(field_name):
                failures.append(f"provenance.{field_name} is required")

    if isinstance(catalog, dict):
        catalog_version = provenance.get("catalog_version")
        if catalog_version is not None and catalog_version != catalog.get("catalog_version"):
            failures.append(
                f"provenance.catalog_version {catalog_version!r} does not match the canonical "
                f"catalog's own {catalog.get('catalog_version')!r}"
            )

    releases = artifact.get("releases", [])
    if not isinstance(releases, list):
        failures.append("releases must be an array")
        releases = []
    release_ids: set[Any] = set()
    for release in releases:
        if not isinstance(release, dict):
            failures.append("release entry must be an object")
            continue
        if set(release) != _RELEASE_KEYS:
            failures.append(
                f"release {release.get('release_id')} has unexpected keys: {sorted(release)}"
            )
            continue
        release_ids.add(release.get("release_id"))

    artists = artifact.get("artists", [])
    if not isinstance(artists, list):
        failures.append("artists must be an array")
        artists = []
    artist_ids = {a.get("artist_id") for a in artists if isinstance(a, dict)}

    albums = artifact.get("albums", [])
    if not isinstance(albums, list):
        failures.append("albums must be an array")
        albums = []
    for album in albums:
        if not isinstance(album, dict):
            failures.append("album entry must be an object")
            continue
        main_release_id = album.get("main_release_id")
        if not isinstance(main_release_id, int) or main_release_id <= 0:
            failures.append(f"album {album.get('id')} has an invalid main_release_id")

    paths = artifact.get("paths", [])
    if not isinstance(paths, list):
        failures.append("paths must be an array")
        paths = []
    for path in paths:
        if not isinstance(path, dict):
            failures.append("path entry must be an object")
            continue
        for hop in path.get("hops", []):
            if not isinstance(hop, dict):
                failures.append(f"path {path.get('id')} hop must be an object")
                continue
            if hop.get("release_id") not in release_ids:
                failures.append(f"path {path.get('id')} references an unpublished release")
            if hop.get("artist_a_id") not in artist_ids or hop.get("artist_b_id") not in artist_ids:
                failures.append(f"path {path.get('id')} references an unpublished artist")

    failures.extend(f"artifact must not have a 'seed' key ({p})" for p in _seed_key_paths(artifact))
    failures.extend(_privacy_failures(artifact, name="artifact"))

    return failures
